### ISSUE TITLE
The API no longer allows local username lookups without user authentication

### ISSUE THREAD

**Issue by Changaco** [mastodon/mastodon#4588]
Before #4155 it was possible to look up an account by username through `/api/v1/search?q={username}@{domain}`, but now that requires a user access token. Trying to access the information using an application token returns `{"error":"This method requires an authenticated user"}` with a 422 status code. `/api/v1/accounts/search` returns the same response.

Could you bring back the possibility of doing a local Account lookup by username without user authentication?

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [x] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by nightpool** [comment]
Can you elaborate a little bit more about why you want to make this request as an unauthenticated user? On the backend, this just delegates to webfinger resolution anyway, why not just do that instead?

**Comment by Changaco** [comment]
> Can you elaborate a little bit more about why you want to make this request as an unauthenticated user?

Because we would rather not have to ask users to login in order to perform a simple username lookup.

> On the backend, this just delegates to webfinger resolution anyway, why not just do that instead?

Trying to get the information we need from somewhere else would require custom code, it's much better if we can simply get an Account object from the API.

**Comment by nightpool** [comment]
The problem with using any one instance to perform "a simple username lookup" is that you're not going to get accurate information unless you get it from the authoritative instance. So, for example, if I look up @Changaco@mastodon.cloud from cybre.space, then it will only succeed if that account has ever been subscribed too or requested before—otherwise it will just return an empty profile. 

The mastodon API is not really intended to be used as a gateway to the wider fediverse, except by users of that instance.

**Comment by nightpool** [comment]
If you're just looking for local users, then you can use `https://hostname.tld/.well-known/webfinger?resource=acct:user@host`.

**Comment by Changaco** [comment]
> The problem with using any one instance to perform "a simple username lookup" is that you're not going to get accurate information unless you get it from the authoritative instance.

I know that, for each account we were sending our "search" query to the authoritative instance. Note that in my original comment I asked for "the possibility of doing a **local** Account lookup by username without user authentication".

> If you're just looking for local users, then you can use `https://hostname.tld/.well-known/webfinger?resource=acct:user@host`, which is the standards-compliant way of doing a user lookup on mastodon instance.

That doesn't give us the same information as an Account object from the API.

**Comment by nightpool** [comment]
Okay, well, the `search` API is definitely not the right call here then. I'll look into the reason the accounts api isn't public.

**Comment by Gargron** [comment]
The search API is too powerful for this use-case. It handles URLs to statuses and accounts, as well as the username@domain URIs, loading them into the database if they don't already exist. This has potential for abuse (DDoS, loading junk into the db) if you allow it to happen unauthenticated. I can see that you need a lookup-by-username API which would not perform any non-DB operations. It wouldn't be hard to add if I had a clue how to name this method and its params.

**Comment by Changaco** [comment]
Some ideas: `/accounts/lookup?username=:username`, `/accounts/lookup/:username`, `/accounts/@:username`.

**Comment by tateisu** [comment]
- check Webfinger for acct
- check the number in salmon URL

```

https://mstdn.maud.io/.well-known/webfinger?resource=https://mstdn.maud.io/@unarist

(snip)
		,{
			"rel":"salmon"
			,"href":"https://mstdn.maud.io/api/salmon/7337"
		}
(snip)

https://mstdn.maud.io/web/accounts/7337
```

**Comment by tateisu** [comment]
for lookup account ID from user name :
- get https://mstdn.maud.io/.well-known/webfinger?resource=acct:unarist@mstdn.maud.io
- parse the number in salmon URL.
```
		,{
			"rel":"salmon"
			,"href":"https://mstdn.maud.io/api/salmon/7337"
		}
```

But the API for account information and account TL requires access-token.
seems this is not performance issue.

**Comment by trwnh** [comment]
This issue can be closed, the API in question was added in v3.4 as `GET /api/v1/accounts/lookup`
